### ISSUE TITLE
[Suggestion] extend 3.6 with upper snake case convention for global constants

### ISSUE THREAD

**Issue by antspk** [goldbergyoni/nodebestpractices#1043]
It's common in other programming languages to use `UPPER_SNAKE_CASE` variable naming convention for global and static constants, even some of the parts of JavaScript and NodeJs use this convention e.g.
```javascript
// js
Math.PI;
Math.LOG10E;

// node
process.env.NODE_ENV;
```
I would suggest to consider this a best practise as well. So it would be an extension to 3.6 when declaring a global or static constant use `UPPER_SNAKE_CASE` convention e.g:

```typescript
export const PLANKS_CONSTANT: number = 6.62607015e-34;

export class Util {
  public static STATIC_CONSTANT: number = 1; 
} 
```

This suggestion wouldn't change 3.6 practise in any way, just extend it.

**Comment by elektronik2k5** [comment]
I agree and use this convention myself too.

**Comment by rluvaton** [comment]
@antspk Thank you so much! 


I use this convention too...

@goldbergyoni WDYT?

**Comment by goldbergyoni** [comment]
Yo @antspk, welcome!

Trying to refine this a but - Isn't the addition of 'const' makes this unnecessary or relevant only for Eexports? See AirBnb take on this:

https://github.com/airbnb/javascript/#naming--uppercase

**Comment by elektronik2k5** [comment]
I don't see any rationale in AirBnB's guideline. A const is a const, whether it is exported or not. Furthermore, if/when I do wanna export an existing (local) const, why should I need to rename it? What value does that have?

In my opinion this is exactly the type of micromanagement which makes people hate overly specific style guides.
Seems like pure mental overhead for zero value.

**Comment by rluvaton** [comment]
I'm with @elektronik2k5 on this one, if the const variable is inside a function I would not upper case it

**Comment by elektronik2k5** [comment]
Oh, yeah, I agree. Not every immutable binding (aka `const`) requires uppercasing.

**Comment by antspk** [comment]
I'm sorry, I see that there's a need to clarify my suggestion, I was NOT suggesting to always use upper snake case convention for ALL  cases of `const` declaration. I think you could separate variables into 3 groups:

`variables`, `values`, `constants`. For example, [Kotlin](https://www.baeldung.com/kotlin/const-var-and-val-keywords) language does that. So if you look at it that way, for the local scope you would use a `val` keyword, which would not need to be uppercased. However, since the syntax of JS is a little different we can't say that every case of `const` should be uppercased.

I don't know how to put it in specific rule/practice terms, but I feel that there are some cases where developers feel the need to emphasize the meaning of a significant value (constant) in which case they do use `UPPER_SNAKE_CASE` convention shout it out in a way.

You can find many examples of this convention, and It kinda feels like this is an unwritten rule that the developers do that, as is not going to change whether it's considered a best practice or not. 

Once again example taken for nodejs http2 module sources: 
```
const HTTP_STATUS_OK: number;
const HTTP_STATUS_CREATED: number;
const HTTP_STATUS_ACCEPTED: number;
const HTTP_STATUS_NO_CONTENT: number;
 ```

So I guess what I would suggest is to add a note to best practices regarding variable naming, that's it's ok to use a `UPPER_SNAKE_CASE` variable naming convention for some exceptional cases where it feels like it's necessary to emphasize the meaning of a value. Although I get that words like `exceptional` and `feels like` are loose and leave out a lot of space for abusive usage of this convention.

**Comment by JFernando122** [comment]
I think there are some key values that separate the `values` and `constants`.

`values` are defined INSIDE a function and it's value can change in each call (Maybe cause the parameters, or a change in other part or something like that).

`constants`are mostly defined OUTSIDE a function (Not always), also often used in many functions, and it's value barely ever change except in VERY specific circunstances.

An example of a `value` could be the response of the db when fetching an element by id. The value won't change once fetched but if you provide a different id then the value will be different.

And for `constant` could be that you set a min and max value for something and those values almost never change.

**Comment by goldbergyoni** [comment]
@antspk Great explanation. I too 'fill' that, but I can't put it behind explicit words. What you proposed 'necessary to emphasize the meaning of a value' feels a little vague to me. 

Eventually, this guide is all about wording, we should clarify the intent.

Opened a discussion here, hope someone will join

https://twitter.com/nodepractices/status/1426926534049292292

**Comment by antspk** [comment]
I think @JFernando122 is onto something here. If we would think philosophically about a `value` and a `constant` as two separate concepts, what is the difference that separates them?

Quick google search yielded these results: 

> In logic, a logical constant of a language is a symbol that has the same semantic value under every interpretation.
[Logical constant](https://en.wikipedia.org/wiki/Logical_constant)

> The difference between value and constant is that a value specified using constant is available during the configuration phase. 
[AngularJS: Constants vs Values](https://stackoverflow.com/questions/30327651/angularjs-constants-vs-values)

> A constant in logic is a symbol that refers to a specific object.
[Philosophy Index](http://www.philosophy-index.com/logic/terms/constant.php)

So if we take the first statement and take a look at this part: "*has the same semantic value under every interpretation*" in a concept of programming (not only NodeJs actually) we could think of a `constant` as some value that can exist outside a project scope and always have the same meaning, for example, a number of *π* or a gravitational acceleration *g*.

However I think that relying on that rule only, would be limiting, and this statement: "*constant is available during the configuration phase*" suites a little bit more, if we would think about a `constant` as a value that is already known at the moment of a definition (in other languages you could also say: known at a compile-time), then it seems to make sense to emphasize it. 

So for example:
```typescript
// an example of a value
function addTwoValues(firstValue: number, secondValue: number): number {
  // the thirdValue is unknown until the function is executed so it's not a constant
  // thus would not need to use UPPER_SNAKE_CASE convention
  const thirdValue: number = firstValue + secondValue;
  
  return thirdValue;
}

// and example of a constant
const NUMBER_OF_PI: number = 3.14159;

function isEqualToTheNumberOfPi(value: number): boolean {
  // in this case the value of NUMBER_OF_PI is known at the moment of a definition
  // which means that it's a constant and should use UPPER_SNAKE_CASE convention
  return value === NUMBER_OF_PI;
}
```

So in summary we should consider few things: 
1. JavaScript `const` can define both a `constant` and `value` and those two are separate concepts. 
2. We should use `UPPER_SNAKE_CASE ` convention for constant and not values.
3. A constant is a value that is known at the moment of a definition and not runtime.

So it doesn't matter if the constant is exported or not, is a static read-only property of a class or a globally declared const value. If the meaning of a value is known at the moment of a definition one should as if it's a constant and maybe should be emphasized and maybe even moved from outside a function scope.

Maybe someone can add something to this, or correct me if I'm wrong, but I think we're onto something here.

**Comment by antspk** [comment]
I was thinking about it a little bit more and started to wonder if there are cases when someone might want to using convention not only for constants but for variables.
e.g.

```typescript
let CALL_COUNT: number = 0;

function call(): void {
  CALL_COUNT++;
}
```

Couldn't think of a meaningful example, but I can image there's might be cases where it would make sense e.g. incrementing atomic integer or something like that.

If the convention would allow for such cases, then the only thing which could be used to determine if the variable should be named using this convention would be scope at which variable is declared. 
```typescript
 // module scope variables - reason for this would be separate global variables from scopes ones 
 const EXAMPLE_CONST = 1;
 let EXAMPLE_VAR = 2;
 
 // function scope variables
 function exampleFunction(): void {
   const scopedConst = 1;
   let scopedVar = 2;
 }

// class scope static variables = module scope variables
class ExampleClass {
  static readonly STATIC_CONST = 2;
  static STATIC_VAR = 1;
  
  exampleFn(): void {
     ExampleClass.STATIC_VAR++;
  }
}
```

So the rule for this could be:
> Use `SCREAMING_SNAKE_CASE` convention for  global and static class variables to separate them from function scoped variables.
